libs/base/Data/IOArrayNew.idr

```diff
@@ -0,0 +1,82 @@
+module Data.IOArrayNew
+
+import Data.IOArray.Prims
+import Data.List
+
+%default total
+
+export
+record IOArrayNew elem where
+  constructor MkIOArrayNew
+  maxSize : Int
+  content : ArrayData elem
+
+export
+max : IOArrayNew elem -> Int
+max = maxSize
+
+export
+newArray : HasIO io => Int -> elem -> io (IOArrayNew elem)
+newArray size init
+    = pure (MkIOArrayNew size !(primIO (prim__newArray size init)))
+
+export
+writeArray : HasIO io => IOArrayNew elem -> Int -> elem -> io ()
```

Comment:
Can we return a `Bool` stating whether the write was successful rather than silently failing?

Reply:
I would prefer to silently crash on index out of range, no?

Because when `writeArray` returns `False`, what else you can do but crash? Likely you have already checked the index is valid before calling this function.

Reply:
> You could be implementing a dynamic array and decide to use newArrayCopy

Typical pattern for that is to check the size before accessing the array:

```
if index >= size array
  then ensureCapacity ...
  else pure ()

writeArray array index elem
```

Do it on write error is somewhat unusual:

```
False <- writeArray array index elem
  | True => pure ()
ensureCapacity ...
ok <- writeArray index elem
and here we need a crash anyway to self-check
```

Also array access by incorrect index is usually a logic error.

But I agreed that quitely ignoring error it definitely not good. I was just following existing `IOArray` implementation.

Reply:
But `writeArray` is already performing the check so it does not make
much sense in my opinion to also perform it yourself before attempting
the write.

Reply:
So you are OK with imposing the responsibility of checking function result on caller, right?

Like consider the matrix example:

```
record Matrix where
  rows : Int
  cols : Int
  data: IOArray

setElement : Matrix -> Int -> Int -> IO ()
setElement m r c = do
  checkRowIndex m c
  checkColumnIndex m c
  ok <- writeArray (data m) ((rows m) * c + r)
  assert ok -- and here we need to self-check, although we have already checked indexes are correct
```

Or hashmap example:

```
record HashMap where
  buckets : Array Bucket -- power of two length

put : HashMap -> Key -> Value -> IO()
put m k v = do
  let hashCode = hash k
  let bucketIndex = hashCode & (size (buckets m) - 1)
  let bucket = readArray buckets bucketIndex
  -- and here we have to check that bucket is not `Nothing` and crash if it is
  -- although we "know" that index is correct
  -- from the way we constructed buckets and computed the index
```

I think this is another case where practicality clashes with purity. Crashing is convenient, but it is not pure.

Reply:
In `readArray` we return a `Maybe`-value, we don't crash if the index is out of bounds.
I am simply advocating following the existing logic here:
1. bound check in the function
2. return an informative value

Reply:
> In readArray we return a Maybe-value

That is the point of my last comment: we can crash instead of returning `Maybe`, it will be more practical. Because caller will likely have to crash anyway otherwise.

Note in Rust there are two read array (slice of `Vec`) and two write array functions: two operators `[]` which crash on out of bounds, and `.get` and `.get_mut` functions which return `Option` on out of bounds. Maybe we can follow this pattern.

Reply:
Here is a proposition:

1. Keep doing what we are doing (except that we return a more informative value in the write case)
2. Add different functions that take *validated* inputs that are proven to be in range and don't bother doing the extra bound checks in that case

This way users can decide whether they want to either observe the result of their operation or bother doing the bound checks manually beforehand.

Reply:
> I just gave you a scenario in which it makes sense to recover from the error without crashing

Resizable array write example. I think I replied to that, I'll elaborate:
* resize on out of range is quite rare data structure, only JavaScript does that I think, and it is not their best design decision
* this is not how resize is usually implemented
* it is more lines of code (second call is needed)
* second call still needs to check the result of operation, and still have to crash on overflow
* in the most cases people just want to crash on out of range

> Add different functions that take validated inputs that are proven to be in range and don't bother doing the extra bound checks in that case

If we have a function which takes validated input, maybe we should just go all in with dependent types then? I. e. have only a function which takes validated input?

Reply:
Personally, I would much rather have a more descriptive return value than silently failing. I understand that it can get tedious having to deal with `Maybe`s and that it complicates extending things, but it also forces the user to consider+handle what happens when things go wrong, which imo is a good thing. It can get annoying when you know it's fine and the machine doesn't, but that's where proofs come in (I'll get back to that). If people prefer things to crash, then let them crash their program; but have them make the decision instead of something magically doing it for them (or if it absolutely must, have it be prefixed with `unsafe_` or something similar). In general, error-handling is one of the things I find neat about FP/Types: Instead of silently failing (or failing but setting some magic value that you just need to know to check, à la `errno`) there is a clear indication of the possibility of error in the return type, in the form of `Maybe` or `Either` or some custom type which can be more informative, which allows the programmer to be more careful by design. So I think it is better to stick to this pattern/approach as much as possible  : )

Alternatively, this could be an experimental approach which exists alongside the existing implementation, but lives in `contrib`. I think this might be the better approach anyway since:

> Existing `IOArray` need to be kept for a while, because the compiler
> need to be compiled with previous Idris version.

If it breaks being able to compile with a previous version, then it sounds like it is a potentially neat, different approach, but one which needs more testing and might change in the future; hence, put it in `contrib`.

Coming back to proofs:

> If we have a function which takes validated input, maybe we should just go all in with dependent types then? I. e. have only a function which takes validated input?

Maybe that is a better approach. I think it would make sense to have to prove that the indexes are in bounds if you want to guarantee that the read/write is fine. It might be necessary to use `believe_me` in the primitive function to assert that if we get to the primitive function, everything is actually fine. But as long as this is never exposed to the user, that might work?

Reply:
> I understand that it can get tedious having to deal with Maybes and that it complicates extending things, but it also forces the user to consider+handle what happens when things go wrong, which imo is a good thing

OK, user now knows that write or read may be out of bounds. And then what? My understanding is that they will simply insert idris_crash (or even worse, return some meaningless value like zero and leave program in buggy state), and that's it. Big ergonomics loss and no increase in safety.

> if it absolutely must, have it be prefixed with unsafe_ or something similar

It is not that unsafe. Unsafe is data corruption. And assertion crash is just a bug in program. Programs may have bugs, and crash is not the worst of them (in fact, it's much easier to debug failed runtime assertion than logic error).

> If it breaks being able to compile with a previous version, then it sounds like it is a potentially neat, different approach, but one which needs more testing and might change in the future; hence, put it in contrib.

The goal here is to replace `IOArray` with new implementation after the next version bump. It is not an alternative, it is replacement.

>  I think it would make sense to have to prove that the indexes are in bounds if you want to guarantee that the read/write is fine

My only worry that it might make code unnecessary verbose (unergonomic).

I was thinking this patterns:

```
data IndexValid : IOArray a -> Int -> Type where ...

checkIndexValid : (arr : IOArray a) -> (index : Int) -> Maybe (IndexValid arr index)

assertIndexValid : (arr : IOArray a) -> (index : Int) -> IndexValid arr index
assertIndexValid arr index = case checkIndexValid arr index of
  Just valid => valid
  Nothing => idris_crash "uh oh"

readArray : (arr : IOArray a) -> (index : Int) -> {indexValid : IndexValid arr index} -> a

myProgram : HasIO io => IOArray String -> io String
myProgram arr = do
  let _ = assertIndexValid arr 0 -- this line might be unnecessarily verbose
  readArray arr 0
```

So I think my API preference is two functions per read and write:
* crashing
* with proof